src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -397,5 +399,42 @@ public bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo>
 
             return converter;
         }
+
+        public static void ValidateType(Type type, Type? parentClassType, PropertyInfo? propertyInfo)
+        {
+            if (IsInvalidForSerialization(type))
+            {
+                ThrowHelper.ThrowInvalidOperationException_CannotSerializeInvalidType(type, parentClassType, propertyInfo);
+            }
+        }
+
+        public static bool IsInvalidForSerialization(Type type)
+        {
+            return type.IsPointer || IsByRefLike(type) || type.ContainsGenericParameters;
+        }
+
+        public static bool IsByRefLike(Type type)
+        {
+#if BUILDING_INBOX_LIBRARY
+            return type.IsByRefLike;
+#else
+            if (!type.IsValueType)
+            {
+                return false;
+            }
+
+            object[] attributes = type.GetCustomAttributes(inherit: false);
+
+            for (int i = 0; i < attributes.Length; i++)
+            {
+                if (attributes[i].GetType().FullName == "System.Runtime.CompilerServices.IsByRefLikeAttribute")
```

Comment:
When was this attribute introduced?  If it was at the same time as IsByRefLike, any reason not to just use reflection to look for IsByRefLike and invoke it rather than looking for this attribute?

Reply:
I can't see much interaction based on commit history, but using reflection to look for `IsByRefLike` appears to be a correct approach based on CI runs.

"Expose IsByRefLike api" - https://github.com/dotnet/corefx/pull/23659
"Adding System\Runtime\CompilerServices\IsByRefLikeAttribute type" - https://github.com/dotnet/coreclr/pull/11369

<details>
<summary>
Benchmarks
</summary>

This approach provides better performance than looking for the attribute.

|                    Method |       type |         Mean |      Error |     StdDev |    Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |----------- |-------------:|-----------:|-----------:|---------:|--------:|-------:|------:|------:|----------:|
|               UseProperty |      Int32 |     1.085 ns |  0.0474 ns |  0.0507 ns |     1.00 |    0.00 |      - |     - |     - |         - |
|           LookupAttribute |      Int32 | 1,499.229 ns | 29.1261 ns | 27.2445 ns | 1,383.28 |   68.58 | 0.1221 |     - |     - |     512 B |
| UsePropertyWithReflection |      Int32 |   118.795 ns |  0.5778 ns |  0.4825 ns |   109.23 |    5.59 | 0.0057 |     - |     - |      24 B |
|                           |            |              |            |            |          |         |        |       |       |           |
|               UseProperty | Span<Byte> |     1.573 ns |  0.0331 ns |  0.0293 ns |     1.00 |    0.00 |      - |     - |     - |         - |
|           LookupAttribute | Span<Byte> | 6,762.362 ns | 96.4735 ns | 90.2413 ns | 4,301.13 |   95.58 | 0.4730 |     - |     - |    2009 B |
| UsePropertyWithReflection | Span<Byte> |   119.345 ns |  0.2818 ns |  0.2498 ns |    75.90 |    1.46 | 0.0057 |     - |     - |      24 B |
|                           |            |              |            |            |          |         |        |       |       |           |
|               UseProperty |     String |     1.382 ns |  0.0566 ns |  0.1091 ns |     1.00 |    0.00 |      - |     - |     - |         - |
|           LookupAttribute |     String | 3,047.869 ns | 28.7296 ns | 25.4680 ns | 2,271.04 |  172.32 | 0.2251 |     - |     - |     952 B |
| UsePropertyWithReflection |     String |    21.498 ns |  0.2538 ns |  0.2250 ns |    16.02 |    1.28 |      - |     - |     - |         - |

```csharp
using BenchmarkDotNet.Attributes;
using System;
using System.Reflection;

namespace MiscBenchmarks
{
    [MemoryDiagnoser]
    public class IsByRefLike
    {
        [Params(typeof(Span<byte>), typeof(string), typeof(int))]
        public Type type;

        private static readonly PropertyInfo s_isByRefLikePropertyInfo = typeof(Type).GetProperty("IsByRefLike");

        [Benchmark(Baseline = true)]
        public bool UseProperty()
        {
            return type.IsByRefLike;
        }

        [Benchmark]
        public bool LookupAttribute()
        {
            object[] attributes = type.GetCustomAttributes(inherit: false);

            for (int i = 0; i < attributes.Length; i++)
            {
                if (attributes[i].GetType().FullName == "System.Runtime.CompilerServices.IsByRefLikeAttribute")
                {
                    return true;
                }
            }

            return false;
        }

        [Benchmark]
        public bool UsePropertyWithReflection()
        {
            if (!type.IsValueType)
            {
                return false;
            }

            return (bool)(s_isByRefLikePropertyInfo.GetValue(type)!);
        }
    }
}
```
</details>